src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/MstatObjectDumper.cs

```diff
@@ -2,68 +2,71 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.IO;
+using System.Linq;
+using System.Reflection;
 using System.Reflection.Metadata;
 using System.Reflection.Metadata.Ecma335;
+using System.Reflection.PortableExecutable;
 
-using Internal.Text;
 using Internal.TypeSystem;
 
 using ILCompiler.DependencyAnalysis;
+using ILCompiler.DependencyAnalysisFramework;
 
 using ObjectData = ILCompiler.DependencyAnalysis.ObjectNode.ObjectData;
 using AssemblyName = System.Reflection.AssemblyName;
-using System.Collections.Generic;
 
 namespace ILCompiler
 {
     public class MstatObjectDumper : ObjectDumper
     {
-        private const int VersionMajor = 1;
-        private const int VersionMinor = 1;
+        private const int VersionMajor = 2;
+        private const int VersionMinor = 0;
 
         private readonly string _fileName;
-        private readonly TypeSystemMetadataEmitter _emitter;
+        private readonly MstatEmitter _emitter;
 
         private readonly InstructionEncoder _types = new InstructionEncoder(new BlobBuilder());
 
-        private Dictionary<MethodDesc, (string MangledName, int Size, int GcInfoSize)> _methods = new();
+        private readonly BlobBuilder _mangledNames = new BlobBuilder();
+
+        private List<(MethodDesc Method, string MangledName, int Size, int GcInfoSize)> _methods = new();
         private Dictionary<MethodDesc, int> _methodEhInfo = new();
         private Dictionary<string, int> _blobs = new();
 
-        private Utf8StringBuilder _utf8StringBuilder = new Utf8StringBuilder();
-
         public MstatObjectDumper(string fileName, TypeSystemContext context)
         {
             _fileName = fileName;
             var asmName = new AssemblyName(Path.GetFileName(fileName));
             asmName.Version = new Version(VersionMajor, VersionMinor);
-            _emitter = new TypeSystemMetadataEmitter(asmName, context);
+            _emitter = new MstatEmitter(asmName, context);
             _emitter.AllowUseOfAddGlobalMethod();
         }
 
         internal override void Begin()
         {
         }
 
-        protected override void DumpObjectNode(NameMangler mangler, ObjectNode node, ObjectData objectData)
+        protected override void DumpObjectNode(NodeFactory factory, ObjectNode node, ObjectData objectData)
         {
-            string mangledName = null;
```

Comment:
This part of the diff is about which name to emit. We didn't actually emit it because that part was rolled back in #75328, but computing the name was left. I now realized it's the wrong name.

We have two names for a node: one name is the name of the symbol in the object file. The second name is the name of the node in DGML. They're the same 95% of time. The remaining 5% are types - we distinguish between two forms of a type, "constructed" (with a vtable) and "unconstructed" (e.g. type used in a cast). At object writing time they get collapsed into one. But for analysis purposes they're two distinct things. This switches the name to use the one that can be looked up in DGML. The object name can be useful too, but maybe we can just do something hacky like remove trailing ` constructed` suffix to compute it if it's needed in the future.